### PR TITLE
DrivAerML: T_max=40 cosine schedule (between champion and T_max=50)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Fine-grained sweep between T_max=30 (champion) and T_max=50. T_max=40 provides a middle ground.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 40 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-tmax-sweep
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT historia:

## Results

### DrivAerML T_max=40 cosine schedule

| Metric | Baseline (T_max=30, PR #2898) | T_max=40 | Delta |
|--------|-------------------------------|----------|-------|
| **val best `surface_rel_l2_pct`** | **3.997%** @ ep467 | **12.84%** @ ep87 | +221% (WORSE) |

- **Best val `surface_rel_l2_pct`:** 12.84% at epoch 87
- **Test at best checkpoint:** Available in W&B (training still running to completion)
- **W&B run:** `mzwuykvf` (historia/dm-tmax-40, group: dm-tmax-sweep)
- **Peak memory:** ~6207 MiB (GPU 0)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 40 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-tmax-sweep --agent historia --wandb-name "historia/dm-tmax-40"
```

### What happened

**T_max=40 is significantly worse than the T_max=30 champion and is unstable without grad-clip.**

The training showed three distinct phases:

1. **Healthy convergence (epochs 1-47):** Smooth cosine cycling with progressively deeper troughs: 36.0% → 28.0% → 18.6% → 14.4%. Very promising trajectory.

2. **First instability event (epoch 48):** Gradient explosion at cosine restart — grad norm tripled (0.93 → 2.76), metric spiked to 44%. The model recovered over ~30 epochs, eventually pushing to a new best of **12.84% at epoch 87**.

3. **Terminal divergence (epoch 94+):** Second gradient explosion — grad norm spiked to 11.2, then escalated exponentially: 62 → 73 → 84 → 115 → 178 → 226 → 354. By epoch 131, the model was oscillating around 25-30% with grad norms >300, in irreversible divergence. The model could not recover from this second event.

**Key finding:** T_max=40 without grad-clip produces recurring gradient instability events at longer training horizons. The champion T_max=30 config ran 467 epochs without instability and reached 3.997%. The longer cosine period (40 vs 30 steps) appears to accumulate larger parameter perturbations at LR peaks, triggering cascading gradient explosions that eventually become self-sustaining.

**Trough progression:**
- Cycle 1 (ep7): 36.0%
- Cycle 2 (ep20): 28.0%
- Cycle 3 (ep34): 18.6%
- Cycle 4 (ep45): 14.4% ← pre-instability best
- Cycle 5 (ep60): 27.2% ← first instability recovery
- Cycle 6 (ep73): 16.1% ← recovered
- Cycle 7 (ep87): 12.8% ← **all-time best**
- Cycle 8 (ep100): 38.3% ← second instability, terminal
- Cycle 9 (ep113): 19.3% ← partial recovery, then diverged to grad norms >300

### Verdict

**NEGATIVE.** T_max=40 does not beat the T_max=30 champion (12.84% vs 3.997%). The longer cosine period introduces catastrophic gradient instability without grad-clip. T_max=30 remains optimal for DrivAerML.

### Suggested follow-ups

1. **T_max=40 + grad-clip=1.0:** The instability is clearly a gradient explosion issue. Adding grad-clip could allow T_max=40 to benefit from longer cosine periods while preventing divergence.
2. **T_max=35 sweep:** A smaller step from T_max=30 might avoid the instability threshold while potentially finding slightly deeper basins.
3. **The early convergence (epochs 1-47) of T_max=40 was actually faster than T_max=30** in terms of per-epoch improvement. This suggests longer cosine periods CAN accelerate learning but need stabilization (grad-clip) for long training.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3082 because ADVISOR: Closing — T_max=40 is a dead end on DrivAerML without gc. Best val 12.840% at ep87 (221% above 3.997% baseline), then diverged with grad norms reaching 2222. This confirms the T_max sweep findings: T_max=15/20/40/50/100 all diverge without gradient clipping. T_max=30 remains the uniquely stable operating point for DrivAerML. Thanks for the thorough three-phase analysis, historia..

**Comment by tcapelle** [comment]
STUDENT historia:

## Test metrics update

Training completed. Test evaluation at best checkpoint (epoch 87):

| Metric | Value |
|--------|-------|
| **test `surface_rel_l2_pct`** | **13.43%** |
| **val best `surface_rel_l2_pct`** | **12.84%** @ ep87 |

Val-test gap is small (0.59 points), confirming the model generalizes well at the best checkpoint. Final training ran 522 steps total before timeout. The divergence did not affect the best-checkpoint test evaluation.
